### DISCUSSION TITLE
Family modifier with future procider behavior I dont understand

### DISCUSSION THREAD

**Discussion by Wykeless** [rrousselGit/riverpod#2800] [Q&A] [answered]
Hello so I am using a family modifier with a future provider and the state of the page doesn't want to update.

version of sdk/package:
> dart - 3.0.6
> flutter - 3.10.6
> flutter_riverpod - 2.3.6

api_service.dart code:
```
import 'dart:developer';
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:riverpod_tut/models/suggestion.dart';

final apiServiceProvider = Provider<ApiService>((ref) => ApiService());

class ApiService {
  Future<Suggestion> getSuggestion(String id) async {
    try {
      final res = await Dio()
          .get('https://www.boredapi.com/api/activity?participants=$id');
      log('Got Data');
      log(res.toString());

      return Suggestion.fromJson(res.data);
    } catch (e) {
      log(e.toString());
      throw Exception('Error getting suggestion');
    }
  }
}
```

future_provider.dart code:
```
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:riverpod_tut/models/suggestion.dart';
import 'package:riverpod_tut/services/api_service.dart';

final suggestionFutureProvider =
    FutureProvider.autoDispose.family<Suggestion, String>((ref, id) async {
  final apiService = ref.watch(apiServiceProvider);
  return apiService.getSuggestion(id);
});

final class FutureProviderPage extends ConsumerWidget {
  const FutureProviderPage({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final suggestionRef = ref.watch(suggestionFutureProvider('1'));
    return Scaffold(
      appBar: AppBar(),
      body: RefreshIndicator(
        //refrsh only works on lists
        onRefresh: () => ref.refresh(suggestionFutureProvider('4').future),
        child: ListView(
          children: [
            suggestionRef.when(data: (data) {
              return Text(data.activity.toString());
            }, error: (error, _) {
              return Text(error.toString());
            }, loading: () {
              return const CircularProgressIndicator();
            }),
          ],
        ),
      ),
    );
  }
}
```
Without modifier and no "?participants=$id" appended:
Video link: [https://imgur.com/a/WSjDKvj](https://imgur.com/a/WSjDKvj)

api call data:
> [log] Got Data
> [log] {"activity":"Go stargazing","type":"relaxation","participants":1,"price":0,"link":"","key":"8832605","accessibility":0.1}
> [log] Got Data On Refresh
> [log] {"activity":"Clean out your garage","type":"busywork","participants":1,"price":0,"link":"","key":"7023703","accessibility":0.1}

with modifier and "?participants=$id" appended:
Video link: [https://imgur.com/a/fMxEIB3](https://imgur.com/a/fMxEIB3)

> [log] Got Data
> [log] {"activity":"Make bread from scratch","type":"cooking","participants":1,"price":0.2,"link":"","key":"4809815","accessibility":0.2}
> [log] Got Data On Refresh
> [log] {"activity":"Play a volleyball match with some friends","type":"social","participants":4,"price":0,"link":"","key":"4306710","accessibility":0.3}

I don't understand the above behavior maybe there's something wrong i did, or maybe its a bug? Cant anyone help me please.

**Comment by AhmedLSayed9** [comment]
It's a feature.
User can see the old data until the new data is being fetched and it also avoid showing 2 loading indicators which is not a good user experience. It's familiar and you can find it on almost every popular mobile application.

However, if you want to disable it, set `skipLoadingOnRefresh` to false at `.when`

**Reply by Wykeless** [reply]
Oh I see, I assume to show new data I have to add that right?
which I tried now, I added the skipLoadingOnRefresh to the .when function and set it to false but the old data is still being shown.
how do I show the new data?

**Reply by AhmedLSayed9** [reply, marked as answer]
At your example above, you're refreshing the wrong member of your family provider.
You should refresh the same member that you're watching:
```dart
onRefresh: () => ref.refresh(suggestionFutureProvider('1').future),
```
You can also refresh the whole family and keep the loading until a specific member is fetched:
```dart
onRefresh: () {
   ref.invalidate(suggestionFutureProvider);
   return ref.read(suggestionFutureProvider('1').future);
},
```

**Reply by Wykeless** [reply]
Okay I see now that, that does work thank you! but now i have another question why doesnt the code below work but the version with the same string '1' does? does changing the string make it a different member?

```
onRefresh: () => ref.refresh(suggestionFutureProvider('4').future),
```

but changing the string in the second block of code works fine, is that because its refreshing the whole family?

Like the reason for changing the string is to an activity for a specific amount of participants

**Reply by AhmedLSayed9** [reply]
When you use some "x" family parameter, you access a provider specifically associated with that unique "x" parameter (as a Map Key). That's why family parameters should have a consistent ==, or else you'll access a new different provider every time.

Doing:
```dart
onRefresh: () => ref.refresh(suggestionFutureProvider('4').future),
```
Is exactly the same as doing:
```dart
onRefresh: (){
   ref.invalidate(suggestionFutureProvider('4'));
   return ref.read(suggestionFutureProvider('4').future);
},
```
You're invalidating the provider keyed with `4` causing it to dispose and then await the new future result of it. That has nothing to do with the provider keyed with `1` which you're using.

**Reply by Wykeless** [reply]
I see, I think I understand it now, thank you.

**Comment by catalunha** [comment]
> That's why family parameters should have a consistent ==, or else you'll access a new different provider every time.

@AhmedLSayed9 
How would this ==
Could you give an example or link something to documentation about this?

**Reply by AhmedLSayed9** [reply]
Consistent means cached or override == (Equatable).
https://docs-v2.riverpod.dev/docs/concepts/modifiers/family#parameter-restrictions

**Comment by catalunha** [comment]
> Ideally, the parameter should either be a primitive (bool/int/double/String), a constant (providers), or an immutable object that overrides == and hashCode.

I understood. When dealing with a class as a parameter. Right ?

**Reply by AhmedLSayed9** [reply]
When dealing with non-primitive types such as classes and collections (list/map).
